### ISSUE TITLE
[HOLD for payment 2023-12-08] [$500] Web - App marks task as unread even when no changes are done

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#27296]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the app
2. Open task report or raise new task report using any user and any title
3. Open Description
4. Without making any changes, click save and observe that task is marked unread

## Expected Result:
Task should not be marked as unread if there is no change to any field

## Actual Result:
Task is marked as unread even if there is no change to any field

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.68.13
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/1ae7e03f-4b76-4033-9a13-21c9e4b510b0

https://github.com/Expensify/App/assets/93399543/80c39294-2225-4d2c-9daa-42f0ffe0bf87

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1694345875816589

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014012f6817510a0d1</li>
        <li>Upwork Job ID: 1701705833501474816</li>
        <li>Last Price Increase: 2023-10-03</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @dylanexpensify</details>

**Comment by iamuddeshya** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue.
Task on the LHN is marked as unread even if the task is saved with no changes made.

## What is the root cause of that problem?

The `editTaskAndNavigate` function is calling the API even when there is no change in the current data and the updated data of the task. 

## What changes do you think we should make in order to solve the problem?

Before calling the EditTask API.write function, add a if condition that checks if there has been any change made in the previous state vs updated state of the Task. If not then avoid calling the API to update the task.
https://github.com/Expensify/App/blob/54085f9ddc8f20826f83b12f4178bb4c986ad9a2/src/libs/actions/Task.js#L478

## What alternative solutions did you explore? (Optional)
No.

**Comment by iamuddeshya** [comment]
Contributor details ua@hyathi.com
Your Expensify account email: ua@hyathi.com
Upwork Profile Link: https://www.upwork.com/freelancers/~011b2c513d4b7b3536

**Comment by iamuddeshya** [comment]
Contributor details
Your Expensify account email: [ua@hyathi.com](mailto:ua@hyathi.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~011b2c513d4b7b3536

**Comment by dylanexpensify** [comment]
Agree this seems like a bug, but confirming first internally

**Comment by sobitneupane** [comment]
@dylanexpensify I'm uncertain about the rationale behind marking it as 'unread' initially for the user initiating the change. In my opinion, it would be more logical to designate the task report as 'unread' solely for users other than the one initiating the change.

**Comment by dylanexpensify** [comment]
@sobitneupane interesting - so my thinking is if no changes have been made - why would it be unread? Nothing new is there that would require attention. What do you think?

**Comment by kevinksullivan** [comment]
I seem like a duplicate here so I'm going to unassign

**Comment by iamuddeshya** [comment]
> > interesting - so my thinking is if no changes have been made - why would it be unread? Nothing new is there that would require attention. What do you think?
> 
> @dylanexpensify I agree. We should not mark it as unread.
> 
> > I'm uncertain about the rationale behind marking it as 'unread' initially for the user initiating the change. In my opinion, it would be more logical to designate the task report as 'unread' solely for users other than the one initiating the change.
> 
> Even if change is made, do we want to mark it as unread for the user making the change? We should mark it as unread only for the user other than the one who made the change. What do you think?

Although, wether if we keep the "read" or "unread" part aside, we must also bat an eye to not call the API if there is no change being made. This is considered as a best practice to only call the API when there is actually a change in the current state vs the updated needed to be done in the state. 
Explained [here](https://github.com/Expensify/App/issues/27296#issuecomment-1716614728) 

cc - @dylanexpensify

**Comment by sobitneupane** [comment]
@tienifr I think it's okay to notify other users that something has changed in the task.

**Comment by sobitneupane** [comment]
https://github.com/Expensify/App/issues/27296#issuecomment-1730960788

@tienifr I agree with your suggestion.

We can have something like. "Description Updated" , "Title Updated", etc.

**Comment by dylanexpensify** [comment]
Reviewing!

**Comment by sobitneupane** [comment]
@dylanexpensify You might have missed [this comment](https://github.com/Expensify/App/issues/27296#issuecomment-1728996619).

I agree with not marking it unread if no change is made. 

I am also suggesting to not mark unread for the user who is making changes (even if change is made) as he/she is aware of the change and is in the same report.

We can also implement something as [suggested](https://github.com/Expensify/App/issues/27296#issuecomment-1730960788) by @tienifr to notify users what exactly has changed. We have similar implementation in request money page as well.
<img width="1062" alt="Screenshot 2023-10-02 at 21 02 11" src="https://github.com/Expensify/App/assets/25876548/2967182e-d839-4653-9ed7-ecb5140f7bc9">

**Comment by dylanexpensify** [comment]
Sorry! To confirm are we asking if we mark as unread for all parties when this is done? Or just for the initiator?

**Comment by sobitneupane** [comment]
@dylanexpensify I think we should mark it as unread for all parties except initiator.

**Comment by dylanexpensify** [comment]
Ok @sobitneupane, and why? If nothing is changing/no new information?

**Comment by sobitneupane** [comment]
@dylanexpensify, let me clarify:

If nothing is changing, I agree not to show it as 'unread' to all users.

However, if something has changed, I propose showing it as 'unread' to all parties except the initiator. Currently, we mark the task report as 'unread' for the initiator as well. While this may not be directly related to this issue, we can discuss it separately.

**Comment by dylanexpensify** [comment]
ahhh thank you for clarifying @sobitneupane!

>However, if something has changed, I propose showing it as 'unread' to all parties except the initiator. Currently, we mark the task report as 'unread' for the initiator as well. While this may not be directly related to this issue, we can discuss it separately.

I agree with this, but let's bring it up in a new issue

**Comment by sobitneupane** [comment]
@dylanexpensify Sorry for the confusion.

I think we still need to solve the issue of "app marked as unread still if nothing has changed" here. I will review the proposals once you reopen the issue.

**Comment by dylanexpensify** [comment]
Ahh that makes sense, reopened @sobitneupane!

**Comment by sobitneupane** [comment]
Sorry for the delay here. I was OOO. I will try to get to it in the weekend.

**Comment by dylanexpensify** [comment]
Thank you @sobitneupane

**Comment by sobitneupane** [comment]
[Proposal](https://github.com/Expensify/App/issues/27296#issuecomment-1716701310) from @hoangzinh looks good to me. Let's make sure to make the change in all the required fields.

🎀 👀 🎀 C+ reviewed

**Comment by sobitneupane** [comment]
@rayane-djouah The approach is same for all the fields. So, I don't think we can consider your proposal substantially different from the other.

**Comment by sobitneupane** [comment]
@tienifr Your suggestion is a feature request. I think that should be dealt in a different issue as discussed [earlier](https://github.com/Expensify/App/issues/27296#issuecomment-1756041419).

**Comment by stitesExpensify** [comment]
@hoangzinh proposal looks good to me

**Comment by stitesExpensify** [comment]
I think we'll leave it as is for now unless real users report that they are confused

**Comment by dylanexpensify** [comment]
@hoangzinh mind giving us an eta on when pr will be ready?

**Comment by dylanexpensify** [comment]
remove overdue mel

**Comment by dylanexpensify** [comment]
Melvin....

**Comment by dylanexpensify** [comment]
Cleaned them all up, but this one isn't moving the overdue label away despite me being assigned and updating it? 

<img width="1512" alt="Screenshot 2023-11-06 at 10 58 31 AM" src="https://github.com/Expensify/App/assets/60117248/573c047e-9f26-42bd-85db-df75ef7153e6">

**Comment by dylanexpensify** [comment]
Sorry wrong button

**Comment by dylanexpensify** [comment]
Sounds good, thanks for the update! ❤️

**Comment by dylanexpensify** [comment]
@hoangzinh mind giving us an updated ETA?

**Comment by dylanexpensify** [comment]
@sobitneupane bump!

**Comment by sobitneupane** [comment]
There are still some [issues](https://github.com/Expensify/App/pull/30844#discussion_r1401684018) in the PR. The PR is Work in Progress.

**Comment by dylanexpensify** [comment]
payment upcoming!

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [ ] The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/18672
> * [ ] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

https://github.com/Expensify/App/pull/18672#issuecomment-1847056109
> * [ ] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:


> * [ ] Determine if we should create a regression test for this bug.

Yes
> * [ ]  [@hoangzinh / @sobitneupane] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

**Comment by sobitneupane** [comment]
**Regression Test Proposal:**

1. Open an existing task report or create a new task report
2. Press on Task description
3. Save the description without making any change
4. Verify that the Task report is not marked as unread in LHN.
5. Repeat Step 2, 3 and 4 for Task title and Task assignee.

Do we agree 👍 or 👎

**Comment by dylanexpensify** [comment]
Payment summary:
- Reporter: @dhanashree-sawant $50
- Contributor: @hoangzinh $500
- Contributor+: @sobitneupane $500

Please apply!

**Comment by dylanexpensify** [comment]
[New Upwork Job!](https://www.upwork.com/jobs/~0115adb603b7b0a41a)

**Comment by dylanexpensify** [comment]
Please apply! ^

**Comment by dylanexpensify** [comment]
Bump @dhanashree-sawant!

**Comment by dhanashree-sawant** [comment]
Sorry for the delay, I have applied to the job

**Comment by sobitneupane** [comment]
https://github.com/Expensify/App/issues/27296#issuecomment-1849560189

Requested payment on newDot.

**Comment by JmillsExpensify** [comment]
$500 payment approved for @sobitneupane based on [this comment](https://github.com/Expensify/App/issues/27296#issuecomment-1849560189).

**Comment by dylanexpensify** [comment]
nice! Will pay out reporter then close!

**Comment by dylanexpensify** [comment]
@dhanashree-sawant offer sent!

**Comment by dhanashree-sawant** [comment]
Thanks @dylanexpensify, I have accepted the offer.

**Comment by dylanexpensify** [comment]
done!
